subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionState.kt

```diff
@@ -188,15 +189,24 @@ class InstantExecutionState(
 
     private
     fun Encoder.writeRelevantProjectsFor(nodes: List<Node>) {
-        writeCollection(fillTheGapsOf(relevantProjectPathsFor(nodes))) { project ->
+        writeCollection(fillTheGapsOf(relevantProjectsFor(nodes))) { project ->
             writeString(project.path)
             writeFile(project.projectDir)
         }
     }
 
     private
-    fun relevantProjectPathsFor(nodes: List<Node>): List<Project> =
-        nodes.mapNotNullTo(mutableListOf()) { node ->
+    fun relevantProjectsFor(nodes: List<Node>): Sequence<Project> =
+        nodes.asSequence().flatMap { node ->
+            sequence {
+                yieldAll(node.dependencySuccessors)
```

Comment:
This is duplicating the logic for traversing the work graph, which should not be required in order to decide which projects are involved. A better fix would be to figure out why there are nodes that are reachable via these edges but which are not included in the `nodes` collection, which represents all of the nodes to be serialized. Later on we also assume that `nodes` contains all of the nodes to be serialized in order to assign ids to the nodes and serialise them, so presumably that will also fail (either with an exception or silently) in the case we're fixing here.

Reply:
FTR: the serialization would fail silently, https://github.com/gradle/gradle/blob/b2f226289fb2261801679ad0a3e73f17767c2403/subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/serialization/codecs/WorkNodeCodec.kt#L145

If I make it throw when the scheduleNodeId can't be found I can observe failures in the tests for `shouldRunAfter` and `mustRunAfter`. I'll give `WorkNodeCodec` a thorough review.

Reply:
Ok. Nevermind my comment above, the code is correct. Any scheduled `shouldRunAfter` or `mustRunAfter` successors would have already been written at that point and the code guards against writing references to non-scheduled successors.